### PR TITLE
add deeplabv3 plus for reproducibility.

### PR DESCRIPTION
**Pull request by tanzhenyu**
This is to help verify the existing PR for developing DeepLabV3Plus, see https://github.com/keras-team/keras-cv/pull/1128

### CODE REVIEW

File: keras_cv/models/segmentation/deeplab_v3_plus.py [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+    lr_decay = tf.keras.optimizers.schedules.PiecewiseConstantDecay(
+        boundaries=[30000 * 16 / global_batch],
+        values=[base_lr, 0.1 * base_lr],
+    )
+    # backbone = keras_cv.models.ResNet50V2(
+    #     include_rescaling=True, weights="imagenet", include_top=False,
+    #     input_shape=[512, 512, 3]
+    # ).as_backbone()
+    # model = DeepLabV3(classes=21, backbone=backbone, include_rescaling=True)
+    model = DeeplabV3Plus(image_size=512, num_classes=21)
```
Comment (DavidLandup0) [inline keras_cv/models/segmentation/deeplab_v3_plus.py:242]: Looks like the backbone is missing?

Comment (tanzhenyu) [reply keras_cv/models/segmentation/deeplab_v3_plus.py]: @DavidLandup0 Updated

Comment (tanzhenyu) [reply keras_cv/models/segmentation/deeplab_v3_plus.py]: I actually updated with a new training script as well, so now you should be able to reproduce it

Comment (DavidLandup0) [reply keras_cv/models/segmentation/deeplab_v3_plus.py]: Awesome, thanks! I did notice a difference in some longer runs, where mine used to plateau faster than the TB logs, and was wondering what was going on...
Can't wait to update the `basic_training.py` to include these as KPLs!

Comment (tanzhenyu) [reply keras_cv/models/segmentation/deeplab_v3_plus.py]: KPLs is a great feature, makes things much easier!
I do notice that segmentation mostly just use simple resizing, so updating training to only use KPL is good to go. 
However we're rolling back KPL for detection models since it has been consistently trained with lower precision, and we need to figure out the logic difference.

Comment (DavidLandup0) [reply keras_cv/models/segmentation/deeplab_v3_plus.py]: I've also noticed that when migrating the sem seg script fully (initial PR that was dropped), the augmentations seem to have slightly reduced the performance of the model. Visualizing the images and maps seemed to have verified a perfectly fine operation for each of them, but there might be some side effect that we aren't considering.

The runs with only flipping practically always outperformed the rest, which is kind of surprising...

**Comment by DavidLandup0** [comment]
Updated my implementation to include dropout like yours. That was the primary model-level difference.

When running with this training script (with augs), the loss is lower than the one I've been running so far (in the previous PR), but the mIOU is also lower:

```
Epoch 1/100
2125/2125 [==============================] - 626s 278ms/step - loss: 0.6756 - sparse_categorical_crossentropy: 0.6756 - mean_io_u: 0.2756 - sparse_categorical_accuracy: 0.8100 - val_loss: 0.8513 - val_sparse_categorical_crossentropy: 0.8513 - val_mean_io_u: 0.4142 - val_sparse_categorical_accuracy: 0.8105
Epoch 2/100
2125/2125 [==============================] - 576s 271ms/step - loss: 0.4873 - sparse_categorical_crossentropy: 0.4873 - mean_io_u: 0.4113 - sparse_categorical_accuracy: 0.8530 - val_loss: 0.6397 - val_sparse_categorical_crossentropy: 0.6397 - val_mean_io_u: 0.4280 - val_sparse_categorical_accuracy: 0.8332
Epoch 3/100
2125/2125 [==============================] - 576s 271ms/step - loss: 0.4349 - sparse_categorical_crossentropy: 0.4349 - mean_io_u: 0.4635 - sparse_categorical_accuracy: 0.8662 - val_loss: 0.6058 - val_sparse_categorical_crossentropy: 0.6058 - val_mean_io_u: 0.4990 - val_sparse_categorical_accuracy: 0.8351
Epoch 4/100
2125/2125 [==============================] - 576s 271ms/step - loss: 0.4081 - sparse_categorical_crossentropy: 0.4081 - mean_io_u: 0.4869 - sparse_categorical_accuracy: 0.8733 - val_loss: 0.4901 - val_sparse_categorical_crossentropy: 0.4901 - val_mean_io_u: 0.5130 - val_sparse_categorical_accuracy: 0.8588
Epoch 5/100
2125/2125 [==============================] - 578s 272ms/step - loss: 0.3809 - sparse_categorical_crossentropy: 0.3809 - mean_io_u: 0.5110 - sparse_categorical_accuracy: 0.8800 - val_loss: 0.4439 - val_sparse_categorical_crossentropy: 0.4439 - val_mean_io_u: 0.5472 - val_sparse_categorical_accuracy: 0.8551
```

Running with the original script (just flipping), but with dropout shows better validation loss and mIOU:

```
2124/2124 [==============================] - 351s 149ms/step - loss: 0.6966 - sparse_categorical_crossentropy: 0.6966 - mean_io_u: 0.3111 - sparse_categorical_accuracy: 0.8012 - val_loss: 0.8283 - val_sparse_categorical_crossentropy: 0.8283 - val_mean_io_u: 0.4548 - val_sparse_categorical_accuracy: 0.8241
Epoch 2/5
2124/2124 [==============================] - 315s 148ms/step - loss: 0.5032 - sparse_categorical_crossentropy: 0.5032 - mean_io_u: 0.4458 - sparse_categorical_accuracy: 0.8459 - val_loss: 0.5126 - val_sparse_categorical_crossentropy: 0.5126 - val_mean_io_u: 0.4913 - val_sparse_categorical_accuracy: 0.8327
Epoch 3/5
2124/2124 [==============================] - 321s 151ms/step - loss: 0.4548 - sparse_categorical_crossentropy: 0.4548 - mean_io_u: 0.4916 - sparse_categorical_accuracy: 0.8586 - val_loss: 0.5082 - val_sparse_categorical_crossentropy: 0.5082 - val_mean_io_u: 0.5299 - val_sparse_categorical_accuracy: 0.8460
Epoch 4/5
2124/2124 [==============================] - 315s 148ms/step - loss: 0.4145 - sparse_categorical_crossentropy: 0.4145 - mean_io_u: 0.5278 - sparse_categorical_accuracy: 0.8688 - val_loss: 0.4062 - val_sparse_categorical_crossentropy: 0.4062 - val_mean_io_u: 0.5591 - val_sparse_categorical_accuracy: 0.8662
Epoch 5/5
2124/2124 [==============================] - 319s 150ms/step - loss: 0.3813 - sparse_categorical_crossentropy: 0.3813 - mean_io_u: 0.5567 - sparse_categorical_accuracy: 0.8775 - val_loss: 0.4487 - val_sparse_categorical_crossentropy: 0.4487 - val_mean_io_u: 0.5240 - val_sparse_categorical_accuracy: 0.8551
```

I'll do a run with your original model as well a bit later, but think that #1128 might go over 70% mIOU now?

**Comment by DavidLandup0** [comment]
Ran the `resnet_deeplab.DeeplabV3Plus` model from this PR against the supplied script (only changed the image size to 224x224 for faster training and consistency with my runs) and here are the results for the first 5 epochs in contrast to the previous runs:

```
2125/2125 [==============================] - 512s 228ms/step - loss: 0.7640 - sparse_categorical_crossentropy: 0.7601 - mean_io_u: 0.2229 - sparse_categorical_accuracy: 0.7921 - val_loss: 0.5102 - val_sparse_categorical_crossentropy: 0.5062 - val_mean_io_u: 0.4642 - val_sparse_categorical_accuracy: 0.8309
Epoch 2/100
2125/2125 [==============================] - 485s 228ms/step - loss: 0.5789 - sparse_categorical_crossentropy: 0.5748 - mean_io_u: 0.3390 - sparse_categorical_accuracy: 0.8305 - val_loss: 0.8767 - val_sparse_categorical_crossentropy: 0.8725 - val_mean_io_u: 0.4358 - val_sparse_categorical_accuracy: 0.8162
Epoch 3/100
2125/2125 [==============================] - 476s 224ms/step - loss: 0.5144 - sparse_categorical_crossentropy: 0.5102 - mean_io_u: 0.3913 - sparse_categorical_accuracy: 0.8451 - val_loss: 1.6957 - val_sparse_categorical_crossentropy: 1.6913 - val_mean_io_u: 0.4757 - val_sparse_categorical_accuracy: 0.8309
Epoch 4/100
2125/2125 [==============================] - 475s 223ms/step - loss: 0.4616 - sparse_categorical_crossentropy: 0.4571 - mean_io_u: 0.4379 - sparse_categorical_accuracy: 0.8583 - val_loss: 0.6296 - val_sparse_categorical_crossentropy: 0.6250 - val_mean_io_u: 0.4585 - val_sparse_categorical_accuracy: 0.8367
Epoch 5/100
2125/2125 [==============================] - 474s 223ms/step - loss: 0.4372 - sparse_categorical_crossentropy: 0.4326 - mean_io_u: 0.4604 - sparse_categorical_accuracy: 0.8647 - val_loss: 0.6939 - val_sparse_categorical_crossentropy: 0.6892 - val_mean_io_u: 0.5245 - val_sparse_categorical_accuracy: 0.8509
```

Might this be an outdated version and not the one that surpassed 70% mIOU? Going from 68% (V3) to 78% (V3Plus) does sound like a pretty large jump. The difference between V3Plus and V3 shouldn't be that huge if I got the paper right 🤔

**Comment by ianstenbit** [comment]
This seems like it is now stale -- if either of you wants to re-open, feel free to do so!
